java/src/main/native/src/row_conversion.cu

```diff
@@ -187,6 +189,63 @@ struct batch_data {
   std::vector<row_batch> row_batches; // information about each batch such as byte count
 };
 
+rmm::device_uvector<size_type> build_string_row_sizes(table_view const &tbl,
+                                                      size_type fixed_width_and_validity_size,
+                                                      rmm::cuda_stream_view stream) {
+  auto const num_rows = tbl.num_rows();
+  rmm::device_uvector<size_type> d_row_sizes(num_rows, stream);
+
+  std::vector<strings_column_view::offset_iterator> offsets_columns;
+  auto offsets_iter = thrust::make_transform_iterator(
+      tbl.begin(), [](auto const &col) -> strings_column_view::offset_iterator {
+        if (!is_fixed_width(col.type())) {
+          CUDF_EXPECTS(col.type().id() == type_id::STRING, "only string columns are supported!");
+          return strings_column_view(col).offsets_begin();
+        } else {
+          return nullptr;
+        }
+      });
+  std::copy_if(offsets_iter, offsets_iter + tbl.num_columns(), std::back_inserter(offsets_columns),
+               [](auto const offset_ptr) { return offset_ptr != nullptr; });
+
+  auto const num_columns = static_cast<size_type>(offsets_columns.size());
+
+  auto d_offsets_columns = make_device_uvector_async(offsets_columns, stream);
+  auto key_iter = cudf::detail::make_counting_transform_iterator(
+      0, [num_columns] __device__(auto element_idx) { return element_idx / num_columns; });
+  auto data_iter = cudf::detail::make_counting_transform_iterator(
+      0, [d_offsets_columns = d_offsets_columns.data(), num_columns,
+          num_rows] __device__(auto element_idx) {
+        auto const row = element_idx / num_columns;
+        auto const col = element_idx % num_columns;
+
+        return d_offsets_columns[col][row + 1] - d_offsets_columns[col][row];
+      });
+
+  // reduce by key to collapse string sizes for a specific row
+  thrust::reduce_by_key(rmm::exec_policy(stream), key_iter, key_iter + num_columns * num_rows,
+                        data_iter, thrust::make_discard_iterator(), d_row_sizes.begin());
+
+  // transform the row sizes to include fixed width size and alignment
+  thrust::transform(rmm::exec_policy(stream), d_row_sizes.begin(), d_row_sizes.end(),
+                    d_row_sizes.begin(), [fixed_width_and_validity_size] __device__(auto row_size) {
+                      return util::round_up_unsafe(fixed_width_and_validity_size + row_size,
+                                                   JCUDF_ROW_ALIGNMENT);
+                    });
```

Comment:
So one thing you can do here is use a transform output iterator in reduce_by_key so that the results from the reduction get transformed (the `round_up_unsafe`) on the spot instead of doing a separate `thrust::transform` call.

https://thrust.github.io/doc/classthrust_1_1transform__output__iterator.html

Reply:
> use a transform output iterator in reduce_by_key

Brilliant! TIL.

Reply:
Sadly this is going away due to performance improvements from marching this data a column at a time. 😞